### DISCUSSION TITLE
random / seeding / shuffle

### DISCUSSION THREAD

**Discussion by inoas** [gleam-lang/gleam#1902] [Ideas & suggestions]
Caveat: This will rely on default args for good ergonomics

1. `list.shuffle` should take an optional callback that can be used instead statically relying on `float.random(0.0, 1.0)`.
2. Add an optional arg to `float.random` and `int.random` to make it easier to generate slower but better values across the Erlang and Javascript by relying on
   - Erlang: https://www.erlang.org/doc/man/crypto.html#strong_rand_bytes-1
   - JavaScript https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues
3. Add `random_float` and `random_int` functions to `gleam/erlang` and _maybe_ `gleam/javascript` that allow to set a seeder and even constant seed values, for Erlang see: https://www.erlang.org/doc/man/rand.html#description - 
It seems https://www.erlang.org/doc/man/rand.html#seed-1 can use the value of https://www.erlang.org/doc/man/rand.html#export_seed-0 to capture and then return to the previous global state?

**Comment by NicklasXYZ** [comment]
This is kind of the reason why I ended up implementing my own PRNG as it's then also possible ensure consistency across the targets: https://github.com/nicklasxyz/gleam_stats/blob/v1.1.0/src/gleam_stats/generators.gleam#L166-L166

Alg: https://www.pcg-random.org/

Aligning the functionality across the two platforms becomes a bit of a mess otherwise.

**Comment by spicychickensauce** [comment]
In case someone finds this looking for a solution to seed `int.random` and `float.random` without passing around seed values, I leave my solution here. I need it only for the JS target, but for erlang it would be much easier, e.g. just call `rand.seed`. For JS I monkey patched `Math.random`. Of course this is quite hacky, but I think totally acceptable for tests.

```gleam
pub fn foo_test() {
  seed("whatever")
  // randomness is now deterministic
}

@external(javascript, "./helpers.js", "seed")
pub fn seed(str: String) -> Nil
```

```JS
// Mostly copied from: https://stackoverflow.com/a/47593316
export function seed(str) {
  const [a, b, c, d] = cyrb128(str);
  const rand = sfc32(a, b, c, d);

  // advance state a bit
  for (var i = 0; i < 15; i++) rand();

  Math.random = function () {
    return rand();
  };
}

function sfc32(a, b, c, d) {
  return function () {
    a |= 0;
    b |= 0;
    c |= 0;
    d |= 0;
    let t = (((a + b) | 0) + d) | 0;
    d = (d + 1) | 0;
    a = b ^ (b >>> 9);
    b = (c + (c << 3)) | 0;
    c = (c << 21) | (c >>> 11);
    c = (c + t) | 0;
    return (t >>> 0) / 4294967296;
  };
}

function cyrb128(str) {
  let h1 = 1779033703,
    h2 = 3144134277,
    h3 = 1013904242,
    h4 = 2773480762;
  for (let i = 0, k; i < str.length; i++) {
    k = str.charCodeAt(i);
    h1 = h2 ^ Math.imul(h1 ^ k, 597399067);
    h2 = h3 ^ Math.imul(h2 ^ k, 2869860233);
    h3 = h4 ^ Math.imul(h3 ^ k, 951274213);
    h4 = h1 ^ Math.imul(h4 ^ k, 2716044179);
  }
  h1 = Math.imul(h3 ^ (h1 >>> 18), 597399067);
  h2 = Math.imul(h4 ^ (h2 >>> 22), 2869860233);
  h3 = Math.imul(h1 ^ (h3 >>> 17), 951274213);
  h4 = Math.imul(h2 ^ (h4 >>> 19), 2716044179);
  ((h1 ^= h2 ^ h3 ^ h4), (h2 ^= h1), (h3 ^= h1), (h4 ^= h1));
  return [h1 >>> 0, h2 >>> 0, h3 >>> 0, h4 >>> 0];
}
```

**Comment by inoas** [comment]
While I find your solution to be fancy I think this should be done through a separate package that allows seeding or though an extra function that allows seeding (and does not pollute global context while doing so)

**Comment by inoas** [comment]
Such as https://hex.pm/packages/prng / https://github.com/giacomocavalieri/prng .. for those looking for one
